### DISCUSSION TITLE
Windows driver debugging

### DISCUSSION THREAD

**Discussion by tormodvolden** [libusb/libusb#1326] [Q&A]
Microsoft offers a tool to check Windows drivers "Driver Verifier". Especially I/O verification and IRP logging sounds interesting:
- https://learn.microsoft.com/en-us/windows-hardware/drivers/devtest/i-o-verification
- https://learn.microsoft.com/en-us/windows-hardware/drivers/devtest/irp-logging

I have tried this on Windows 10 Pro (in a CMD window with administrative rights):
`verifier /volatile /flags 0x410 /adddriver libusb0.sys
`
Then exercised the driver by running xusb on a device that uses libusb0.sys, and replugging the device.
After this, I see the load/unload count increasing, but no other relevant statistics or errors.
<details> <summary>Details (added another flag in this case)</summary>
C:\WINDOWS\system32>verifier /query

Time Stamp: 10/15/2023 19:44:14.958

Verifier Flags: 0x00080410

  Standard Flags:

    [ ] 0x00000001 Special pool.
    [ ] 0x00000002 Force IRQL checking.
    [ ] 0x00000008 Pool tracking.
    [X] 0x00000010 I/O verification.
    [ ] 0x00000020 Deadlock detection.
    [ ] 0x00000080 DMA checking.
    [ ] 0x00000100 Security checks.
    [ ] 0x00000800 Miscellaneous checks.
    [ ] 0x00020000 DDI compliance checking.

  Additional Flags:

    [ ] 0x00000004 Randomized low resources simulation.
    [ ] 0x00000200 Force pending I/O requests.
    [X] 0x00000400 IRP logging.
    [ ] 0x00002000 Invariant MDL checking for stack.
    [ ] 0x00004000 Invariant MDL checking for driver.
    [ ] 0x00008000 Power framework delay fuzzing.
    [ ] 0x00010000 Port/miniport interface checking.
    [ ] 0x00040000 Systematic low resources simulation.
    [X] 0x00080000 DDI compliance checking (additional).
    [ ] 0x00200000 NDIS/WIFI verification.
    [ ] 0x00800000 Kernel synchronization delay fuzzing.
    [ ] 0x01000000 VM switch verification.
    [ ] 0x02000000 Code integrity checks.

    [X] Indicates flag is enabled.

  Verifier Statistics Summary

    Raise IRQLs:                                     0
    Acquire Spin Locks:                              0
    Synchronize Executions:                          0
    Trims:                                           0

    Pool Allocations Attempted:                9294189
    Pool Allocations Succeeded:                9294189
    Pool Allocations Succeeded SpecialPool:          0
    Pool Allocations With No Tag:                    0
    Pool Allocations Not Tracked:                    0
    Pool Allocations Failed:                         0
    Pool Allocations Failed Deliberately:            0

  Driver Verification List

    MODULE: libusb0.sys (load: 4 / unload: 3)

      Pool Allocation Statistics: ( NonPaged / Paged )

        Current Pool Allocations:  (        0 /        0 )
        Current Pool Bytes:        (        0 /        0 )
        Peak Pool Allocations:     (        0 /        0 )
        Peak Pool Bytes:           (        0 /        0 )
        Contiguous Memory Bytes:              0
        Peak Contiguous Memory Bytes:         0

</details>

`dc2wmiparser.exe` mentions VerifierIrpDetails but doesn't show any log entries.
<details>
DC2WMIParser - Verifier Active Test WMI Parser V1.0
GenerateNSName is called with strNSPrefix:=EnumerateInstances.
GenerateNSName was called with pPreFix:=EnumerateInstances.
DC2WMIParser - CoInitializeEx call worked.
Connecting to  WMI server at root\WMI...

Connected with 
DC2WMIParser - Calling VerifierIrpDetails.
DC2WMIParser - VerifierIrpDetails the call to SysAllocString worked.
DC2WMIParser - VerifierIrpDetails the call to SysAllocString worked for MSVerifierIrpLogInformation.
DC2WMIParser - VerifierIrpDetails the call to pIWbemServices->CreateInstanceEnum return HRESULT:=0.
DC2WMIParser - VerifierIrpDetails the call to pIWbemServices->CreateInstanceEnum worked and hr:=WBEM_NO_ERROR.
DC2WMIParser - VerifierIrpDetails the call to pEnum->Next returned hresult:=1.
DC2WMIParser - The call to VerifierIrpDetails worked and returned status:=0.
</details>

(I have also tried without /volatile and rebooting the computer instead.)
It seems @sonatique shares my experience.

Anyone having a clue about this? @dontech, have you used verifier.exe on libusb0.sys?

**Comment by tormodvolden** [comment]
I had more luck with [IRPMon](https://github.com/MartinDrab/IRPMon). (If you install it, uncheck the Server install options unless you want to do remote monitoring)
See example (text) log: [irpmonlog2.log](https://github.com/libusb/libusb/files/12913191/irpmonlog2.log)

**Comment by mcuee** [comment]
@dontech
Please help to share your experiences as well. Thanks.

**Comment by mcuee** [comment]
I am mainly using the debug version of libusb and libusbK and then DebugView to get the debug log.
https://learn.microsoft.com/en-us/sysinternals/downloads/debugview

I have never tried to debug a Windows driver myself. The official tool is WinDBG.
https://learn.microsoft.com/en-us/windows-hardware/drivers/debugger/getting-started-with-windows-debugging
